rs/system_api/src/lib.rs

```
@@ -3523,6 +3535,155 @@ impl SystemApi for SystemApiImpl {
         result
     }
 
+    fn ic0_cost_call(
+        &self,
+        method_name_size: u64,
+        payload_size: u64,
+        dst: usize,
+        heap: &mut [u8],
+    ) -> HypervisorResult<()> {
+        let execution_mode =
+            WasmExecutionMode::from_is_wasm64(self.sandbox_safe_system_state.is_wasm64_execution);
+        let cost = self
+            .sandbox_safe_system_state
+            .cycles_account_manager
+            .xnet_call_total_fee((method_name_size + payload_size).into(), execution_mode);
```

If the caller provides `u64::MAX` for both arguments, this sum will overflow and the call will return a wrong cost. However, the method name of an actual call is limited to 10'000 bytes. So the input `u64::MAX` for `method_name_size` is "wrong", so I would accept this behaviour.

---

Actually, `payload_size = u64::MAX - 1` is not a valid input, they only go up to 2MB. 
While the limits may change, the are very unlikely to ever multiply to even close to `u64::MAX`. 
I would argue that giving a bad result for bad inputs is inevitable, and it does not make it more correct if we e.g. saturate: The result would still be wrong.

---

The `xnet` prefix is used in the cycles account manager for all inter-canister calls.

---

As far as I understand, we have a limit for `ic0_call` payload of 2 MB, which is fine. But in the current implementation, user can call `ic0_cost_call(method_name: "asd", payload_size: u64::MAX - 1, dst);` there will be integer overflow, and the user will not understand what happened. What I propose is to add a check for the `payload_size` parameter of  `ic0_cost_call` that in the case `payload_size` is larger than the payload limit(in this case 2MB), the system returns an error with a message.

---

I believe that we should make consistent behavior of `ic0_call` and `ic0_cost_call`. Taking that into account in the solution you proposed, if the user calls `ic0_cost_call` with payload_size 3MB, it will receive the response with some cost, if, after that, the user calls `ic0_call` with a payload of 3MB call will fail because it is above the payload limit.

---

I think the cost endpoint should only care about the potential cost, not about other things like limits or whether e.g. the payload is malformed or anything like that. I mean, ic0.call_perform can also fail because of many other reasons, we don't want to reflect them in the cost endpoint. Imagine e.g. an empty method name. That would give a legal cost, but is not a legal argument to call_perform. 

In other words, I propose: 
- If call_perform _would_ fail because we don't have enough cycles, then cost_call should help detecting that
- If call_perform would fail for any other reason (e.g. too large inputs), that has nothing to do with the cost, so cost_call will be agnostic about it. 

If we want to return an error code for the cost_call endpoint, we have to specify it (or extend the one from ic0.call_perform), and I don't think it's worth specifying "your inputs to cost calculation were too large".

---

If the payload is malformed that is not information that we have at the moment of calling `ic0_cost_call` so we cannot know what the payload will be hence we should not do anything. But we know `payload_size` in the moment of calling `ic0_cost_call` so with that information if the `payload_size` is greater than the limit will most certainly tell us that the call will not succeed, hence in my opinion that function that should predict the cost of such call(`ic0_cost_call`) returns some cost when we know that such call will not succeed.

Regarding the empty method name, I also wanted to mention that the check for that should be added for the same reason as `payload_size`, we know a call with such a method name will fail so it does not make sense that function that should predict the cost of such call(`ic0_cost_call`) returns some cost when in the moment of calling it we know such call will not succeed.

---

IMO this would not separate concerns properly and leak error states/-handling of `call_perform` into `cost_call`. 

Regarding the second paragraph: E.g. I think it's not wrong to say that e.g. `cost_call(0, 0) = 0`, instead of `cost_call(0, 0) = error` (or some base cost instead of 0, but that's not the point). It's just about the potential cost, nothing else should matter. 

Ok, we can discuss it afk.

---

Discussed offline, we concluded that there is not strong opinion on either side, so a point will be raised on spec PR to get another opinion.

---

https://github.com/dfinity/portal/pull/4110/files#r1958196545